File: backend/app/core/security.py
import hashlib
import secrets
import uuid
from datetime import datetime, timedelta, timezone
from typing import Any, Optional

from jose import JWTError, jwt
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from fastapi import HTTPException, status

from app.core.config import get_settings

settings = get_settings()


def create_access_token(
    data: dict[str, Any], expires_delta: Optional[timedelta] = None
) -> str:
    """Create JWT access token with jti claim."""
    to_encode = data.copy()
    expire = datetime.now(timezone.utc) + (
        expires_delta or settings.JWT_ACCESS_TOKEN_EXPIRE_DELTA
    )
    to_encode.update({
        "exp": expire,
        "jti": str(uuid.uuid4()),
        "type": "access",
        "iat": datetime.now(timezone.utc),
    })
    return jwt.encode(to_encode, settings.SECRET_KEY, algorithm=settings.JWT_ALGORITHM)


def create_refresh_token(data: dict[str, Any]) -> str:
    """Create JWT refresh token."""
    to_encode = data.copy()
    expire = datetime.now(timezone.utc) + settings.JWT_REFRESH_TOKEN_EXPIRE_DELTA
    to_encode.update({
        "exp": expire,
        "jti": str(uuid.uuid4()),
        "type": "refresh",
        "iat": datetime.now(timezone.utc),
    })
    return jwt.encode(to_encode, settings.SECRET_KEY, algorithm=settings.JWT_ALGORITHM)


def decode_token(token: str) -> dict[str, Any]:
    """Decode and validate JWT. Raises 401 on failure."""
    try:
        payload = jwt.decode(
            token, settings.SECRET_KEY, algorithms=[settings.JWT_ALGORITHM]
        )
        return payload
    except JWTError as e:
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail=f"Invalid token: {str(e)}",
            headers={"WWW-Authenticate": "Bearer"},
        )


def get_token_blacklist_key(jti: str) -> str:
    """Redis key for JWT blacklist entry."""
    return f"jwt_blacklist:{jti}"


def hash_token(token: str) -> str:
    """SHA256 hash of API token."""
    return hashlib.sha256(token.encode()).hexdigest()


def verify_token_hash(token: str, token_hash: str) -> bool:
    """Verify token matches stored hash."""
    return hash_token(token) == token_hash


def generate_api_token() -> str:
    """Generate sk-company-xxxxx format token."""
    return f"sk-company-{secrets.token_hex(20)}"


def encrypt_value(plaintext: str, key: str) -> str:
    """AES-256-GCM encrypt. Returns base64(nonce + ciphertext + tag)."""
    import base64
    key_bytes = key.encode("utf-8")
    # Ensure 32-byte key via SHA256
    aes_key = hashlib.sha256(key_bytes).digest()
    aesgcm = AESGCM(aes_key)
    nonce = secrets.token_bytes(12)
    ciphertext = aesgcm.encrypt(nonce, plaintext.encode("utf-8"), None)
    return base64.b64encode(nonce + ciphertext).decode("utf-8")


def decrypt_value(ciphertext_b64: str, key: str) -> str:
    """AES-256-GCM decrypt."""
    import base64
    key_bytes = key.encode("utf-8")
    aes_key = hashlib.sha256(key_bytes).digest()
    aesgcm = AESGCM(aes_key)
    data = base64.b64decode(ciphertext_b64)
    nonce = data[:12]
    ciphertext = data[12:]
    plaintext = aesgcm.decrypt(nonce, ciphertext, None)
    return plaintext.decode("utf-8")


def mask_sensitive(value: str, show_prefix: int = 3, show_suffix: int = 4) -> str:
    """Mask middle of string: show first N and last M chars."""
    if len(value) <= show_prefix + show_suffix + 4:
        return value
    return value[:show_prefix] + "*" * (len(value) - show_prefix - show_suffix) + value[-show_suffix:]

File: backend-rs/src/security.rs
//! Security utilities: JWT create/decode, API token generation, AES encryption.
//!
//! Ported from `backend/app/core/security.py`

use std::time::{SystemTime, UNIX_EPOCH};

use aes_gcm::{
    aead::{Aead, KeyInit},
    Aes256Gcm, Nonce,
};
use base64::Engine;
use base64::engine::general_purpose::STANDARD as BASE64;
use jsonwebtoken::{decode, encode, DecodingKey, EncodingKey, Header, Validation};
use rand::Rng;
use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};
use uuid::Uuid;

use crate::config::AppConfig;

// ============================================================
// JWT
// ============================================================

/// Claims in our JWT tokens
#[derive(Debug, Serialize, Deserialize)]
pub struct Claims {
    /// Subject (user id)
    pub sub: String,
    /// JWT ID (for blacklisting)
    pub jti: String,
    /// Token type: "access" or "refresh"
    #[serde(rename = "type")]
    pub token_type: String,
    /// Expiration (unix timestamp)
    pub exp: usize,
    /// Issued at (unix timestamp)
    pub iat: usize,
    /// User role (only for access tokens)
    pub role: Option<String>,
}

/// Create a JWT access token (30 min default)
pub fn create_access_token(
    data: serde_json::Value,
    config: &AppConfig,
) -> Result<String, String> {
    let now = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map_err(|e| format!("Time error: {}", e))?
        .as_secs() as usize;

    let expires_in = config.jwt_access_token_expire_minutes as usize * 60;

    let sub = data["sub"].as_str().unwrap_or_default().to_string();
    let role = data["role"].as_str().map(|s| s.to_string());

    let claims = Claims {
        sub,
        jti: Uuid::new_v4().to_string(),
        token_type: "access".to_string(),
        exp: now + expires_in,
        iat: now,
        role,
    };

    encode(
        &Header::default(),
        &claims,
        &EncodingKey::from_secret(config.secret_key.as_bytes()),
    )
    .map_err(|e| format!("JWT encode error: {}", e))
}

/// Create a JWT refresh token (7 days default)
pub fn create_refresh_token(
    data: serde_json::Value,
    config: &AppConfig,
) -> Result<String, String> {
    let now = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map_err(|e| format!("Time error: {}", e))?
        .as_secs() as usize;

    let expires_in = config.jwt_refresh_token_expire_days as usize * 86400;

    let sub = data["sub"].as_str().unwrap_or_default().to_string();
    let role = data["role"].as_str().map(|s| s.to_string());

    let claims = Claims {
        sub,
        jti: Uuid::new_v4().to_string(),
        token_type: "refresh".to_string(),
        exp: now + expires_in,
        iat: now,
        role,
    };

    encode(
        &Header::default(),
        &claims,
        &EncodingKey::from_secret(config.secret_key.as_bytes()),
    )
    .map_err(|e| format!("JWT encode error: {}", e))
}

/// Decode and validate a JWT token. Returns the claims.
pub fn decode_token(token: &str, config: &AppConfig) -> Result<Claims, String> {
    let token_data = decode::<Claims>(
        token,
        &DecodingKey::from_secret(config.secret_key.as_bytes()),
        &Validation::default(),
    )
    .map_err(|e| format!("Invalid token: {}", e))?;

    Ok(token_data.claims)
}

/// Redis key for JWT blacklist entry
pub fn get_token_blacklist_key(jti: &str) -> String {
    format!("jwt_blacklist:{}", jti)
}

// ============================================================
// API Token generation & hashing
// ============================================================

/// Generate an API token in the format `sk-company-{40 hex chars}`
pub fn generate_api_token() -> String {
    let mut rng = rand::thread_rng();
    let random_bytes: [u8; 20] = rng.gen();
    let hex_str = hex::encode(random_bytes);
    format!("sk-company-{}", hex_str)
}

/// SHA256 hash of a token
pub fn hash_token(token: &str) -> String {
    let mut hasher = Sha256::new();
    hasher.update(token.as_bytes());
    hex::encode(hasher.finalize())
}

/// Verify a token against its stored hash
pub fn verify_token_hash(token: &str, token_hash: &str) -> bool {
    hash_token(token) == token_hash
}

// ============================================================
// AES-256-GCM encryption/decryption (matching Python implementation)
// ============================================================

/// AES-256-GCM encrypt. Returns base64(nonce + ciphertext).
/// Uses SHA256 to derive a 32-byte key from the encryption_key string.
pub fn encrypt_value(plaintext: &str, encryption_key: &str) -> Result<String, String> {
    // Derive 32-byte key via SHA256
    let mut hasher = Sha256::new();
    hasher.update(encryption_key.as_bytes());
    let key_bytes = hasher.finalize();

    let key = aes_gcm::Key::<Aes256Gcm>::from_slice(&key_bytes);
    let cipher = Aes256Gcm::new(key);

    let nonce_bytes: [u8; 12] = rand::thread_rng().gen();
    let nonce = Nonce::from_slice(&nonce_bytes);

    let ciphertext = cipher
        .encrypt(nonce, plaintext.as_bytes())
        .map_err(|e| format!("Encrypt error: {:?}", e))?;

    // Concatenate nonce + ciphertext and base64 encode
    let mut result = nonce_bytes.to_vec();
    result.extend_from_slice(&ciphertext);

    Ok(BASE64.encode(&result))
}

/// AES-256-GCM decrypt. Input is base64(nonce + ciphertext).
pub fn decrypt_value(ciphertext_b64: &str, encryption_key: &str) -> Result<String, String> {
    let cipher_bytes = BASE64
        .decode(ciphertext_b64)
        .map_err(|e| format!("Base64 decode error: {}", e))?;

    if cipher_bytes.len() < 12 {
        return Err("Ciphertext too short".to_string());
    }

    let nonce = &cipher_bytes[..12];
    let ciphertext = &cipher_bytes[12..];

    // Derive 32-byte key via SHA256
    let mut hasher = Sha256::new();
    hasher.update(encryption_key.as_bytes());
    let key_bytes = hasher.finalize();

    let key = aes_gcm::Key::<Aes256Gcm>::from_slice(&key_bytes);
    let cipher = Aes256Gcm::new(key);
    let nonce = Nonce::from_slice(nonce);

    let plaintext = cipher
        .decrypt(nonce, ciphertext)
        .map_err(|e| format!("Decrypt error: {:?}", e))?;

    String::from_utf8(plaintext).map_err(|e| format!("UTF-8 error: {}", e))
}

// ============================================================
// Utility
// ============================================================

/// Mask middle of string: show first N and last M chars
pub fn mask_sensitive(value: &str, show_prefix: usize, show_suffix: usize) -> String {
    let len = value.len();
    if len <= show_prefix + show_suffix + 4 {
        return value.to_string();
    }
    let prefix = &value[..show_prefix];
    let suffix = &value[len - show_suffix..];
    let stars = "*".repeat(len - show_prefix - show_suffix);
    format!("{}{}{}", prefix, stars, suffix)
}
